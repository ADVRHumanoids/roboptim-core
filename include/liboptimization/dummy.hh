// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Implementation of the dummy module (always fail).
 */

#ifndef OPTIMIZATION_DUMMY_HH
# define OPTIMIZATION_DUMMY_HH
# include <liboptimization/solver.hh>

namespace optimization
{
  /// Dummy solver which always fails.
  class DummySolver : public Solver<Function, Function>
  {
  public:
    /// Define parent's type.
    typedef Solver<Function, Function> parent_t;

    /// Constructor.
    explicit DummySolver (const problem_t&) throw ();
    /// Destructor.
    virtual ~DummySolver () throw ();

    /// Implement the solve algorithm.
    virtual void solve () throw ();
  };

}; // end of namespace optimization

#endif //! OPTIMIZATION_DUMMY_HH
