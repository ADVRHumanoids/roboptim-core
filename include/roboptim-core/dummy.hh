// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \brief Implementation of the dummy module (always fail).
 */

#ifndef ROBOPTIM_CORE_DUMMY_HH
# define ROBOPTIM_CORE_DUMMY_HH
# include <roboptim-core/solver.hh>

namespace roboptim
{
  /// Dummy solver which always fails.
  class DummySolver : public Solver<Function, const Function*>
  {
  public:
    /// Define parent's type.
    typedef Solver<Function, const Function*> parent_t;

    /// Constructor.
    explicit DummySolver (const problem_t&) throw ();
    /// Destructor.
    virtual ~DummySolver () throw ();

    /// Implement the solve algorithm.
    virtual void solve () throw ();
  };

} // end of namespace roboptim

#endif //! ROBOPTIM_CORE_DUMMY_HH
