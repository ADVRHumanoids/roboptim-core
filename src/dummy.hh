// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \file src/dummy.hh
 *
 * \brief Implementation of the dummy module (always fail).
 */

#ifndef OPTIMIZATION_DUMMY_HH
# define OPTIMIZATION_DUMMY_HH
# include <solver.hh>

namespace optimization
{
  /// Dummy solver which always fails.
  class DummySolver : public Solver
  {
  public:
    /// Constructor.
    explicit DummySolver (const Problem&) throw ();
    /// Destructor.
    virtual ~DummySolver () throw ();
    /// Return the default solver error.
    virtual result_t getMinimum () throw ();
  };

} // end of namespace optimization

#endif //! OPTIMIZATION_DUMMY_HH
