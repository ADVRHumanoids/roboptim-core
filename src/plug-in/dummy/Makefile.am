include $(top_srcdir)/build-aux/init.mk

lib_LTLIBRARIES = liboptimization-dummy.la

liboptimization_dummy_la_SOURCES = dummy.cc
liboptimization_dummy_la_LDFLAGS = -module
