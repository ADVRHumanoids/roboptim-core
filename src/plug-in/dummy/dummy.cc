// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \file dummy.cc
 *
 * \brief Implementation of the dummy module (always fail).
 */

#include <solver.hh>

namespace optimization
{

  class DummySolver : public Solver
  {
  public:
    DummySolver (const function_t& fct) throw ()
      : Solver (fct)
    {
    }

    virtual ~DummySolver () throw ()
    {
    }

    virtual result_t
    getMinimum () throw ()
    {
      return result_t (SolverError ());
    }
  };

} // end of namespace optimization


extern "C"
{
  optimization::Solver*
  create (const optimization::DummySolver::function_t& arg)
  {
    return new optimization::DummySolver (arg);
  }

  void
  destroy (optimization::Solver* p)
  {
    delete p;
  }

  unsigned
  version ()
  {
    return optimization::Solver::PLUGIN_INTERFACE_VERSION;
  }
}
