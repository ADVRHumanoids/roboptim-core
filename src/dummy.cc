// Copyright (C) 2009 by Thomas Moulard, FIXME.
//
// This file is part of the liboptimization.
//
// liboptimization is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// liboptimization is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with liboptimization.  If not, see <http://www.gnu.org/licenses/>.


/**
 * \file src/dummy.cc
 *
 * \brief Implementation of the dummy module (always fail).
 */

#include "dummy.hh"

namespace optimization
{
  DummySolver::DummySolver (function_t fct,
                            size_type n,
                            gradient_t g,
                            hessian_t h,
                            jacobian_t j) throw ()
    : Solver (fct, n, g, h, j)
  {
    result_ = SolverError ("The dummy solver always fail.");
  }

  DummySolver::~DummySolver () throw ()
  {
  }

  DummySolver::result_t
  DummySolver::getMinimum () throw ()
  {
    return result_;
  }

} // end of namespace optimization
