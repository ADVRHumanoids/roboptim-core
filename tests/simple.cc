// Copyright (C) 2009 by Thomas Moulard, AIST, CNRS, INRIA.
//
// This file is part of the roboptim.
//
// roboptim is free software: you can redistribute it and/or modify
// it under the terms of the GNU Lesser General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// roboptim is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU Lesser General Public License for more details.
//
// You should have received a copy of the GNU Lesser General Public License
// along with roboptim.  If not, see <http://www.gnu.org/licenses/>.

#include <iostream>
#include <boost/variant/get.hpp>

#include "common.hh"
#include <roboptim-core/dummy.hh>

using namespace roboptim;

typedef DummySolver solver_t;

struct F : public Function
{
  F () : Function (4, 1)
  {}

  virtual vector_t operator () (const vector_t& x) const throw ()
  {
    vector_t res (m);
    res (0) = x[0] * x[3] * (x[0] + x[1] + x[2]) + x[3];
    return res;
  }

  // No gradient, hessian.
};

int run_test ()
{
  F f;
  DummySolver::problem_t pb (f);

  solver_t solver (pb);
  solver_t::result_t res = solver.minimum ();
  boost::get<SolverError> (res);

  std::cout << pb << std::endl
            << "---" << std::endl
            << solver << std::endl;

  // Try to get the minimum from a GenericSolver*
  GenericSolver* gs = &solver;
  std::cout << boost::get<SolverError> (gs->minimum ()).what ()
            << std::endl;

  return 0;
}

GENERATE_TEST ()
