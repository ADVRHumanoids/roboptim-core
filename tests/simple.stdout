Problem:
  a * d * (a + b + c) + d (not derivable)
  Argument's bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
  Argument's scales: 1, 1, 1, 1
  No constraints.
  No starting point.
  Infinity value (for all functions): inf
Problem:
  a * d * (a + b + c) + d (not derivable)
  Argument's bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
  Argument's scales: 1, 1, 1, 1
  Number of constraints: 1
  Constraint 0
      a * d * (a + b + c) + d (not derivable)
      Bounds: (0, 5)
      Scales: 3.5
      Initial value: [1](0)
      
  Starting point: [4](0,0,0,0)
  Starting value: [1](0)
  Infinity value (for all functions): inf
---
Solver:
  Problem:
    a * d * (a + b + c) + d (not derivable)
    Argument's bounds: (-inf, inf), (-inf, inf), (-inf, inf), (-inf, inf)
    Argument's scales: 1, 1, 1, 1
    Number of constraints: 1
    Constraint 0
        a * d * (a + b + c) + d (not derivable)
        Bounds: (0, 5)
        Scales: 3.5
        Initial value: [1](0)
        
    Starting point: [4](0,0,0,0)
    Starting value: [1](0)
    Infinity value (for all functions): inf
  Parameters:
    dummy-parameter (dummy parameter): 42
    dummy-parameter2 (yet another dummy parameter): 3
    dummy-parameter3 (just a dummy key): ...and a dummy value!
    
The dummy solver always fail.
Discrete point: 2.3 (cnt: 1)
Discrete point: 2.8 (cnt: 2)

Discrete point: 2 (cnt: 1)
Discrete point: 2.1 (cnt: 2)
Discrete point: 2.2 (cnt: 3)
Discrete point: 2.3 (cnt: 4)
Discrete point: 2.4 (cnt: 5)
Discrete point: 2.5 (cnt: 6)
Discrete point: 2.6 (cnt: 7)
Discrete point: 2.7 (cnt: 8)
Discrete point: 2.8 (cnt: 9)
Discrete point: 2.9 (cnt: 10)
Discrete point: 3 (cnt: 11)

Discrete point: 0.8 (cnt: 1)
Discrete point: 1.8 (cnt: 2)
Discrete point: 2.8 (cnt: 3)
Discrete point: 3.8 (cnt: 4)
Discrete point: 4.8 (cnt: 5)
Discrete point: 5.8 (cnt: 6)
Discrete point: 6.8 (cnt: 7)
Discrete point: 7.8 (cnt: 8)
Discrete point: 8.8 (cnt: 9)
Discrete point: 9.8 (cnt: 10)
Discrete point: 10.8 (cnt: 11)

